homeassistant/components/plaato/__init__.py

```diff
@@ -59,32 +75,116 @@
+            )
 
+    entry.add_update_listener(async_reload_entry)
     return True
 
 
-async def async_unload_entry(hass, entry):
+def setup_webhook(hass: HomeAssistant, entry: ConfigEntry):
+    """Init webhook based on config entry."""
+    if entry.data[CONF_WEBHOOK_ID] is not None:
+        webhook_id = entry.data[CONF_WEBHOOK_ID]
+        device_name = entry.data[CONF_DEVICE_NAME]
+        hass.components.webhook.async_register(
+            DOMAIN, f"{DOMAIN}.{device_name}", webhook_id, handle_webhook
+        )
+    else:
+        raise InvalidStateError
+
+
+async def async_setup_coordinator(hass: HomeAssistant, entry: ConfigEntry):
+    """Init auth token based on config entry."""
+    auth_token = entry.data.get(CONF_TOKEN)
+    device_type = entry.data.get(CONF_DEVICE_TYPE)
+
+    coordinator = PlaatoCoordinator(hass, auth_token, device_type)
+    await coordinator.async_refresh()
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    for platform in PLATFORMS:
+        if entry.options.get(platform, True):
+            coordinator.platforms.append(platform)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Unload a config entry."""
-    hass.components.webhook.async_unregister(entry.data[CONF_WEBHOOK_ID])
-    hass.data[SENSOR_DATA_KEY]()
+    use_webhook = entry.data.get(CONF_USE_WEBHOOK)
+
+    if use_webhook:
+        return await async_unload_webhook(hass, entry)
+
+    return await async_unload_coordinator(hass, entry)
 
-    await hass.config_entries.async_forward_entry_unload(entry, SENSOR)
-    return True
+
+async def async_unload_webhook(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload webhook based entry."""
+    if entry.data[CONF_WEBHOOK_ID] is not None:
+        hass.components.webhook.async_unregister(entry.data[CONF_WEBHOOK_ID])
+
+    unloaded = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
+    if unloaded:
+        hass.data[SENSOR_DATA_KEY]()
+
+    return unloaded
+
+
+async def async_unload_coordinator(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload auth token based entry."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    unloaded = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+                if platform in coordinator.platforms
+            ]
+        )
+    )
+    if unloaded:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unloaded
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Reload config entry."""
+    await async_unload_entry(hass, entry)
+    await async_setup_entry(hass, entry)
 
```

Comment:
```suggestion
```

You shouldn't need this as this is what
`await hass.config_entries.async_reload(entry.entry_id)` does